README
Introduction:
=============
---------------------
python-haystack is an heap analysis framework, focused on classic 
C structure matching.

The first class of algorithms gives the ability to search for known
structures in a live process's memory, or in a memory dump.

The second class of algorithms aims at giving a reverse engineering look
at a memory dump, focusing on reconstruction, classification of classic
C structures from memory. Heap analysis. Dynamic types definition.

How to get a memory dump:
=========================
---------------------
While technically you could use a third party tool, haystack actually 
need memory mapping information to work with. 
So there is a dumping tool included::

$ sudo haystack-dump dump <pid> dumps/myssh.dump

You can easily reproduce the format of the dump, its a folder/archive 
containing each memory map in a separate file :
- content in a file named after it's start/end addresses ( 0x000700000-0x000800000 )
- 'mappings' file containing memory mappings metadata.  ( mappings )


Search for known structures:
============================
---------------------
You need the python definition/ctypes structures of the known 
structures.
An example would be sslsnoop, which provide python ctypes structures for
openssl and openssh structures.

Quick info:
 This demonstrate the ability to brute-force the search 
 of a known structure, based on fields types assumptions or constraints. 

The magic is performed in the model.py module.
The constraints are applied on the python ctypes structures by the 
'expectedValues' static field.

Command line example: 
---------------------

For example, this will dump the session_state structures + pointed 
children structures as an python object that we can play with.
Lets assume we have an ssh client or server as pid 4042::

  $ sudo haystack --pid 4042 sslsnoop.ctypes_openssh.session_state search > instance.pickled
  $ sudo haystack --pid 4042 sslsnoop.ctypes_openssh.session_state refresh 0xb8b70d18 > instance.pickled
  $ sudo haystack --pid xxxx <your ctypes Structure> search > instance.pickled

or (as root)::

  >>> import haystack
  >>> state_it = haystack.search_process('sslsnoop.ctypes_openssh.session_state', 4042, hint=0)
  >>> state = state_it.next()[0][0]
  >>> state.receive_context.evp.app_data.aes_ctx.rd_key
  '\xcc\xeaM#\xbd# \xc1\x89\xf5\xaa\xb7\xc6f!\x91\xfe\x17\xcc\x97C4\xecV\xca\xc1F\xe1\x0c\xa7gp\xa0\x92\x9di\xe3\xa6q?)g7\xde%\xc0P\xae\x1e\xc1yV\xfdg\x08i\xd4\x00?\xb7\xf1\xc0o\x19\xaci\xad\xf7Q\x0e\xa5\x9e\x85\x0e\x9a)t\xce\xf507\x8f\xa9ef\x81\x0c\xfb\xe3\x8f\x96\xd2\x97Ac\xe2\x94t1\xed\xf2\xf5=\x16\x11z\xab\xc4\x86;\xc8&6\x9c\xc6\xa9\xc4i\xfb\xbf\xd5\x13P{S(\x98]\x82\xda\x8aDF\xb3q\xfb\x93\xa0!\x80\xc0\x88\xb9\xdd]\x8cK\xfe\x1b?:\x05\x88\x9f\x1b\x85H\x17\xa2X\x9b\xb6!\xac\x80\x89\x1b\xa9\x08\x16\x00,@\x01\xa2t\xd5LC\x7f\xdc\xa3\xfc5\xc3\xc1\xa8\xb3\xdcul\xfca\xfaG\xd7\x85r\xf1\x92\x93\xd5cn\xefa\xa5\x88l\xd0#\xfb2\x00H\xdc%\xed^\xdf\xa1\x86yFK\xaf\xcd\xe7)\xb2\xdd\xcb\xd1\xa8\xad\xb0\xdf\xb1\xb8E'

and that was the session key of the receive stream.

Graphic example :
-----------------

There is also an attempt at a Graphical GUI ( Qt4 )
Dump the process, then you can open it in the GUI::

$ haystack-gui # ( and Ctrl-O , click click)
$ haystack-gui --dumpname dumps/myssh.dump

You can the search a structure from the heap of that memory mapping.
You have to import your extensions before that to have them listed in 
the search dialog.
( try sslsnoop.ctypes_openssh ) 

Tip:
 As this is a beta version, sslsnoop is hard-imported in the GUI. 
 You should have it installed.


python script interpreter example:
----------------------------------

::

>>> import haystack
>>> pid = 4042
>>> haystack.findStruct( pid , sslsnoop.ctypes_openssh.session_state)
>>> haystack.findStruct( pid , ctypes_example.big_struct)


Extensibility:
--------------

It's easy to add new structures. Its basically the ctypes definition of 
C structures that should be done following the next 4 steps :

#) Your class must extend haystack.model.LoadableMembersStructure.
#) You must give your class a completed _fields_ (with one _ ), like 
   all ctypes.Structure 
#) Optional - You can add an expectedValues dict() to your ctype 
       classes to add some constraints.
#) Optional - You can override isValid and loadMembers to implement
       advanced constraints validation.
#) call model.registerModule(sys.modules[__name__])

Easy 'creation':
  use h2xml and xml2py from ctypeslib to generate a python module from
  a C header.

Advanced use:
  You can override methods isValid and loadMembers to implements 
  advanced data loading and constraints validation.
  ( see sslsnoop for loading cipher structures from void pointers )

The global algorithm :
  #) The ctypes structure is mapped at the first offset of the memory 
     mapping.
  #) The method loadMembers is called.
  #) The method isValid is called on self.
  #) A validation test is done for each members, constraints and 
     memory space validity (pointers) are tested.
     The validation does not recurse.
  #) Each members is then 'loaded' to local space. 
     If the value is a pointer or a model.LoadableMembersStructure type, it's 
     recursively Loaded. ( and validated).
     If the recursive loading fails, the calls fails. bye-bye.
  #) If all contraints are respected, we have a match.
  #) Move to see next offset, goto 1)


Heap analysis / Memory Reverser / Memory forensics:
===================================================

Quick info: 
 This tool parse the heap for allocator structures, pointers
 values, small integers and text (ascii/utf).
 Given all the previous information, it can extract instances 
 and helps you in classifying and defining structures types.

::

|    usage: haystack-reverser [-h] [--debug]
|                             dumpname
|                             {instances,typemap,group,parent,graph,show,makesig,clean}
|                             ...
|
|    Several tools to reverse engineer structures on the heap.
|
|    positional arguments:
|      dumpname              Source memory dump by haystack.
|      {instances,typemap,group,parent,graph,show,makesig,clean}
|                            sub-command help
|        instances           List all structures instances with virtual address,
|                            member types guess and info.
|        typemap             Try to reverse generic types from instances'
|                            similarities.
|        group               Show structure instances groups by size and signature.
|        parent              Print the parent structures pointing to the structure
|                            located at this address.
|        graph               DISABLED - Show sorted structure instances groups by
|                            size and signature in a graph.
|        show                Show one structure instance.
|        makesig             Create a simple signature file of the heap - NULL,
|                            POINTERS, OTHER VALUES.
|        clean               Clean the memory dump from cached info.
|
|    optional arguments:
|      -h, --help            show this help message and exit
|      --debug               Debug mode on.


Command line example:
--------------------
This will create several files in the folder containing <yourdumpname>::

$ python haystack-reverse instances <yourdumpname>

  The most insteresting one being the <yourdumpname>.headers_values.py that
  gives you an ctypes listing of all found structures, with gestimates
  on fields types.

  A <yourdumpname>.gexf file is also produced to help you visualize 
  instances links. It gets messy for any kind of serious application.


Show ordered list of structures, by similarities::

$ python haystack-reverse show <yourdumpname>

Show only structures of size 324::

$ python haystack-reverse show --size 324 <yourdumpname>


Write to file an attempt to reversed the original types hierachy::

$ python haystack-reverse typemap <yourdumpname>


Extension examples :
====================
@ see sslsnoop in the Pypi repo. openssl and nss structures are generated.

@ see ctypes-kernel on my github. Linux kernel structure are generated from a build kernel tree. (VMM is abitch)


Pseudo Example for extension :
==============================
|from haystack.model import LoadableMembersStructure, RangeValue, NotNull
|
|class OpenSSLStruct(LoadableMembersStructure):
|  pass
|
|class RSA(OpenSSLStruct):
|  ''' rsa/rsa.h '''
|  _fields_ = [
|  ("pad",  ctypes.c_int), 
|  ("version",  ctypes.c_long),
|  ("meth",ctypes.POINTER(BIGNUM)),#const RSA_METHOD *meth;
|  ("engine",ctypes.POINTER(ENGINE)),#ENGINE *engine;
|  ('n', ctypes.POINTER(BIGNUM) ), ## still in ssh memap
|  ('e', ctypes.POINTER(BIGNUM) ), ## still in ssh memap
|  ('d', ctypes.POINTER(BIGNUM) ), ## still in ssh memap
|  ('p', ctypes.POINTER(BIGNUM) ), ## still in ssh memap
|  ('q', ctypes.POINTER(BIGNUM) ), ## still in ssh memap
|  ('dmp1', ctypes.POINTER(BIGNUM) ),
|  ('dmq1', ctypes.POINTER(BIGNUM) ),
|  ('iqmp', ctypes.POINTER(BIGNUM) ),
|  ("ex_data", CRYPTO_EX_DATA ),
|  ("references", ctypes.c_int),
|  ("flags", ctypes.c_int),
|  ("_method_mod_n", ctypes.POINTER(BN_MONT_CTX) ),
|  ("_method_mod_p", ctypes.POINTER(BN_MONT_CTX) ),
|  ("_method_mod_q", ctypes.POINTER(BN_MONT_CTX) ),
|  ("bignum_data",ctypes.POINTER(ctypes.c_ubyte)), ## moue c_char_p ou POINTER(c_char) ?
|  ("blinding",ctypes.POINTER(BIGNUM)),#BN_BLINDING *blinding;
|  ("mt_blinding",ctypes.POINTER(BIGNUM))#BN_BLINDING *mt_blinding;
|  ]
|  expectedValues={
|    "pad": [0], 
|    "version": [0], 
|    "references": RangeValue(0,0xfff),
|    "n": [NotNull],
|    "e": [NotNull],
|    "d": [NotNull],
|    "p": [NotNull],
|    "q": [NotNull],
|    "dmp1": [NotNull],
|    "dmq1": [NotNull],
|    "iqmp": [NotNull]
|  }
|  def loadMembers(self, mappings, maxDepth):
|    print 'example'
|    if not LoadableMembersStructure.loadMembers(self, mappings, maxDepth):
|      log.debug('RSA not loaded')
|      return False
|    return True
|
|# register to haystack
|model.registerModule(sys.modules[__name__])
|
|#EOF


not so FAQ :
------------

What does it do ?:
------------------
The basic functionnality is to search in a process' memory maps for a 
specific C Structures.
The extended reverse engineering functionnality aims at reversing 
structures from memory/heap analysis.

How do it knows that the structures is valid ? :
------------------------------------------------
You add some constraints ( expectedValues ) on the fields. Pointers are also a good start.
 
Where does the idea comes from ? :
-----------------------------------
http://www.hsc.fr/ressources/breves/passe-partout.html.fr originally.
since I started in March 2011, I have uncovered several other related 
previous work.
Most of them are in the docs/ folder.
Other related work are mona.py from Immunity, some other Madiant stuff...
In a nutshell, this is probably not an original idea. But yet, I could 
not find a operational standalone lib for live memory extraction for sslsnoop, soo....


What are the dependencies ? :
----------------------------
python-ptrace on linux
winappdbg on win32
several others...


CHANGES.txt
Ongoing:
 - stabilisation of search
 - refactorisation of reverse

Changes to architecture.txt:
 - IMemoryHandler: there is a process memory concept (dump, live, vol, rekall) that need a handler.
 - IMemoryLoader: a memory loader is a helper class reads the dump and creates a IMemoryHandler instance
    + figures out the ITargetPlatform information (cpu, wordsize)
        + the ITargetPlatform wraps a CTypesProxy instance
    + figures out the list of IMemoryMapping
    + creates a IMemoryHandler with both previous element
 - the IMemoryHandler instance is the one way access to:
    + the list of IMemoryMapping
    + the CTypesProxy ctypes proxy object for this MemoryHandler host
    + the IHeapWalker helpers
 - the memory l

2017-06:
 - Python 3 compatibility finalized
 - using pkg_resources entry points for CLI, memory mapping loader
 - setup pkg_resources entry points for HEAP parser
 - Support frida as memory mappings loader technique
 - add 'read the docs' documentation in .rst format
 - normalize argument parsing around an URL format

2017-05:
 - Start porting to python3
 - extract extensions (reverse, gui) and docs to other git repo

2016-01:
 - Support Minidump format (Windows standard dump format with all memory options)
 - haystack-find-heap.py script to help debug Windows heap
 - 90% support of LFH and LAL on winXP 32 and 64 bits.
 - 90% support of LFH on win7 32 and 64 bits.
 - 95% support of UCR.
 - Support of mixed x32/x64 Windows Heap processes

2015-10:
 - cleaning up the reverse API
 - Fix sigsegv bug source in reset_mappings()
 - Fix the double linked reverser
 - Cleaned up the reverse/structure module
 - Added a fun constraints heuristics on reversed types's field.
 - is zeus in c30000.headers_values.py ?


2015-09:
 - Working on Volatility dump reader
 - working a WinXP chunk iterator
 - adding PEB to winXP list of know structures
 - Working volatility plugins in repo https://github.com/trolldbois/volatility_plugins
 - winheap to work with heap at w.e. offset, not with heap aligned on mapping.start.
 - fixed winXP x86 heapwalker

2015-08:
 - Fixes for travis in setup.py
 - Ran 225 tests in 519.059s
OK (skipped=7, expected failures=4)
 - Profiling... Ran 225 tests in 160.300s
 - heaps are not necessarly at offset 0 in mapping
 - more work on Windows XP heap based on the Zeus vmem extract of a process
 - cleaned the memory handler to stop being a big bag of everything
 - remove reverse's context from core haystack
 - added a script to find heaps in Windows dumps
 - start standardizing test dumps in a test file inventory
Ran 240 tests in 119.036s
FAILED (failures=2, errors=9, skipped=11, expected failures=3)
 - model is an instance.


2015-07:
 - Ran 212 tests in 58.233s
FAILED (failures=1, errors=106, skipped=6, expected failures=1)
 - move constraints to config flat file instead of python code
 - refactor to ctypes proxy class in instances instead of module.
 - design with abc ready interfaces
 - clean TU state, 0 errors
 - constraints config is case sensitive
 - fixed win7heapfinder and walker
 - adjusted basicmodel to use memeory_handler properly.
 - Ran 213 tests in 149.355s
FAILED (failures=2, errors=31, skipped=8, expected failures=2)
 - fixed the search API to a simple one
 - fixed the haystack search cli to the new API
 - Ran 213 tests in 169.886s
FAILED (errors=22, skipped=6, expected failures=2)
 - pretty much refactored haytack core and search in reverse.
 - Ran 229 tests in 347.116s
FAILED (errors=7, skipped=7, expected failures=3)
 - cleaned search API, with matching TU
 - the constraints model and listmodel are enforced by an MVC model
 - Ran 224 tests in 317.901s
FAILED (failures=2, errors=13, skipped=7, expected failures=3)


2015-06:
 - keep track of changes in this file
 - use travis-ci to compile
 - fix makefile in tests/ and sudo & virtualenv play together nicely
 - FIX OverflowError: Python int too large to convert to C long on ptrace file.seek - ignore vsyscall mmap that is in the higher mmap reagion
    + this is a temporary bugfix.
 - add memory dumps test case to a dropbox
 - fix the heap structures used by haystack for win7.
    + Constraints: Ignore HEAP_ENTRY SubSegmentCode pointer as it is encoded.
    + fix some ctypeslib bugs
 - fix the generation process for HEAP structures. ctypeslib and clang do not like circular dependencies.
 - fix the name generation in ctypeslib
 - autopep8 all
 - fix all TU to pass
 - fix installation requirements
 - fix test run
 - Script the generation of profiles



haystack/cli.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

from __future__ import print_function

import argparse
import logging
import os
import sys
import time

import pkg_resources

try:
    from urllib.parse import urlparse
except ImportError:
    from urlparse import urlparse

from haystack import argparse_utils
from haystack import basicmodel
from haystack import constraints
from haystack.search import api

log = logging.getLogger('cli')

# the description of the function
SEARCH_DESC = 'Search for instance of a record_type in the allocated memory of a process. '
SHOW_DESC = 'Cast the bytes at this address into a record_type. '
WATCH_DESC = 'Cast the bytes at this address into a record_type and refresh regularly. '
DUMP_DESC = 'Extract the process dump from the OS memory dump in haystack format. '

# some dumptype constants
DUMPTYPE_BASE = 'haystack'
DUMPTYPE_VOLATILITY = 'volatility'
DUMPTYPE_REKALL = 'rekall'
DUMPTYPE_LIVE = 'live'
DUMPTYPE_MINIDUMP = 'minidump'
DUMPTYPE_FRIDA = 'frida'

# from urlparse import urlparse
# >>> o = urlparse('http://www.cwi.nl:80/%7Eguido/Python.html')
# ParseResult(scheme='http', netloc='www.cwi.nl:80', path='/%7Eguido/Python.html',
# URL_SCHEMES = {'dir': DUMPTYPE_BASE,
#                'volatility': DUMPTYPE_VOLATILITY,
#                'rekall': DUMPTYPE_REKALL,
#                'live': DUMPTYPE_LIVE,
#                'dmp': DUMPTYPE_MINIDUMP,
#                'frida': DUMPTYPE_FRIDA}

SUPPORTED_DUMP_URI = {}
# populate SUPPORTED_DUMP_URI
for entry_point in pkg_resources.iter_entry_points("haystack.mappings_loader"):
    SUPPORTED_DUMP_URI[entry_point.name] = entry_point.resolve()


def url(u):
    """Validates the argument is an url"""
    # dont populate SUPPORTED_DUMP_URI here, otherwise API wont have it.
    _url = urlparse(u)
    scheme = _url.scheme.lower()
    if scheme not in SUPPORTED_DUMP_URI.keys():
        raise argparse.ArgumentTypeError("Target type {s}:// not supported".format(s=scheme))
    path = _url.path
    # be nice with relative path
    if _url.netloc.startswith('~'):
        path = os.path.expanduser(os.path.sep.join([_url.netloc, path]))
        _url = urlparse("%s://%s" % (scheme, path))
    if _url.netloc.startswith('.'):
        path = os.path.abspath(os.path.sep.join([_url.netloc, path]))
        _url = urlparse("%s://%s" % (scheme, path))
    if scheme in ['volatility', 'rekall']:
        path = _url.path.split(':')[0]
    if scheme in ['dir', 'volatility', 'rekall', 'dmp']:
        if not os.path.exists(path):
            raise argparse.ArgumentTypeError("Target {p} does not exists".format(p=path))
        # see url.netloc for host name, frida ? live ?
    return _url


# the description of the dump type
DUMPTYPE_BASE_DESC = 'The process dump is a folder produced by a haystack-dump script.'
DUMPTYPE_VOL_DESC = 'The process dump is a volatility OS dump. The PID is the targeted process.'
DUMPTYPE_REKALL_DESC = 'The process dump is a rekall OS dump. The PID is the targeted process.'
DUMPTYPE_LIVE_DESC = 'The PID must be a running process.'
DUMPTYPE_MINIDUMP_DESC = 'The process dump is a Minidump (MDMP) process dump.'


class HaystackError(Exception):
    pass


def make_memory_handler(opts):
    dumptype = opts.target.scheme.lower()
    if dumptype not in SUPPORTED_DUMP_URI.keys():
        raise TypeError('dump type has no case support. %s' % dumptype)
    loader = SUPPORTED_DUMP_URI[dumptype](opts)
    return loader.make_memory_handler()


def get_output(memory_handler, results, rtype):
    if rtype == 'string':
        ret = api.output_to_string(memory_handler, results)
    elif rtype == 'python':
        # useful in interactive mode
        ret = api.output_to_python(memory_handler, results)
    elif rtype == 'json':
        ret = api.output_to_json(memory_handler, results)
    elif rtype == 'pickled':
        ret = api.output_to_pickle(memory_handler, results)
    else:
        raise ValueError('unknown output format')
    return ret


def dump_process(opts):
    """ Extract the process dump from the OS memory dump in haystack format. """
    if opts.dumptype == DUMPTYPE_VOLATILITY:
        pass
    elif opts.dumptype == DUMPTYPE_REKALL:
        from haystack.mappings import rek
        rek.rekall_dump_to_haystack(opts.dump_filename, opts.pid, opts.output_folder_name)
    return


def search_cmdline(args):
    """ Search for instance of a record_type in the allocated memory of a process. """
    # get the memory handler adequate for the type requested
    memory_handler = make_memory_handler(args)
    # try to load constraints
    my_constraints = None
    if args.constraints_file:
        handler = constraints.ConstraintsConfigHandler()
        my_constraints = handler.read(args.constraints_file.name)
    # get the python record type
    modulename, sep, classname = args.record_type_name.rpartition('.')
    _module = None
    try:
        _module = memory_handler.get_model().import_module(modulename)
    except ImportError as e:
        log.error('sys.path is %s', sys.path)
        raise e
    record_type = getattr(_module, classname)
    # do the search
    results = api.search_record(memory_handler, record_type, my_constraints, extended_search=args.extended)
    # output handling
    try:
        ret = get_output(memory_handler, results, args.output)
        # print output on stdout
        print(ret)
    except Exception as e:
        log.error(e)
    finally:
        if args.interactive:
            print('results are local variable "results"')
            import code
            code.interact(local=locals())
    return


def show_cmdline(args):
    """Cast the bytes at this address into a record_type. """
    # we need an int
    memory_address = args.address
    # get the memory handler adequate for the type requested
    memory_handler = make_memory_handler(args)
    # check the validity of the address
    heap = memory_handler.is_valid_address_value(memory_address)
    if not heap:
        log.error("the address is not accessible in the memoryMap")
        raise ValueError("the address is not accessible in the memoryMap")
    # get the structure name
    modulename, sep, classname = args.record_type_name.rpartition('.')
    _module = None
    try:
        _module = memory_handler.get_model().import_module(modulename)
    except ImportError as e:
        log.error('sys.path is %s', sys.path)
        raise e
    record_type = getattr(_module, classname)
    # load the record
    result = api.load_record(memory_handler, record_type, memory_address)
    results = [result]
    # validate if required
    validation = None
    if args.constraints_file:
        handler = constraints.ConstraintsConfigHandler()
        my_constraints = handler.read(args.constraints_file.name)
        validation = api.validate_record(memory_handler, result[0], my_constraints)
    # output handling
    ret = None
    try:
        ret = get_output(memory_handler, results, args.output)
        # print output on stdout
        print(ret)
        if args.constraints_file:
            print('Validated', validation)
    except Exception as e:
        log.error(e)
    finally:
        if args.interactive:
            print('results are local variable "results"')
            import code
            code.interact(local=locals())
    return


def check_varname_for_type(memory_handler, varname, struct_type):
    done = []
    st = struct_type
    model = memory_handler.get_model()
    ctypes = memory_handler.get_target_platform().get_target_ctypes()
    for v in varname:
        if not hasattr(st, v):
            fields = ["%s: %s" % (n, t) for n, t in basicmodel.get_fields(st)]
            log.error('(%s.)%s does not exists in type %s\n\t%s', '.'.join(done), v, st, '\n\t'.join(fields))
            return False
        st = st._get_field_type(v)
        if ctypes.is_pointer_type(st):  # accept pointers
            st = model.get_subtype(st)
        done.append(v)
    return True


def get_varname_value(varname, instance):
    done = []
    var = instance
    for v in varname:
        var = getattr(var, v)
        done.append(v)
    return '%s = \n%s' % ('.'.join(done), var)


def watch(args):
    """Cast the bytes at this address into a record_type and refresh regularly. """
    memory_address = args.addr
    refresh = args.refresh_rate
    varname = args.varname
    # get the memory handler adequate for the type requested
    memory_handler = make_memory_handler(args)
    # check the validity of the address
    heap = memory_handler.is_valid_address_value(memory_address)
    if not heap:
        log.error("the address is not accessible in the memoryMap")
        raise ValueError("the address is not accessible in the memoryMap")
    # get the structure name
    modulename, sep, classname = args.record_type_name.rpartition('.')
    _module = None
    try:
        _module = memory_handler.get_model().import_module(modulename)
    except ImportError as e:
        log.error('sys.path is %s', sys.path)
        raise e
    record_type = getattr(_module, classname)
    # verify target fieldcompliance
    if varname is not None:
        varname = varname.split('.')
        if not check_varname_for_type(memory_handler, varname, record_type):
            return False
    # load the record
    result = api.load_record(memory_handler, record_type, memory_address)
    results = [result]
    # output handling
    output = api.output_to_python(memory_handler, results)
    # _get_output(memory_handler, results, rtype):
    # Conflicts with varname
    py_obj = output[0][0]
    # print pyObj
    # print as asked every n secs.
    while True:
        # clear terminal
        print(chr(27) + "[2J")
        #
        if varname is None:
            print(py_obj)
        else:
            print(get_varname_value(varname, py_obj))

        if refresh == 0:
            break
        time.sleep(refresh)
        result = api.load_record(memory_handler, record_type, memory_address)
        results = [result]
        # output handling
        output = api.output_to_python(memory_handler, results)
        py_obj = output[0][0]


def base_argparser(program_name, description):
    """ Base options shared by all console scripts """
    rootparser = argparse.ArgumentParser(prog=program_name, description=description)
    verbosity = rootparser.add_mutually_exclusive_group(required=False)
    verbosity.add_argument('--debug', dest='debug', action='store_true', help='Set verbosity to DEBUG')
    verbosity.add_argument('--quiet', dest='quiet', action='store_true', help='Set verbosity to ERROR only')
    rootparser.add_argument('--interactive', dest='interactive', action='store_true',
                            help='drop to python command line after action')
    rootparser.add_argument('--nommap', dest='mmap', action='store_false', help='disable mmap()-ing')
    rootparser.add_argument('--osname', '-n', action='store', default=None, choices=['linux', 'winxp', 'win7'], help='Force a specific OS')
    rootparser.add_argument('--bits', '-b', type=int, action='store', default=None, choices=[32, 64], help='Force a specific word size')
    text = '://, '.join(sorted(SUPPORTED_DUMP_URI.keys())) + '://'
    help_desc = 'target file or process. Supported URL types: %s' % text
    rootparser.add_argument('target', type=url, help=help_desc)
    return rootparser


def search_argparser(search_parser):
    """ Search function options argument parser """
    search_parser.add_argument('record_type_name', type=str,
                               help='Python record type name. Module must be in Python path')
    search_parser.add_argument('--constraints_file', type=argparse.FileType('r'),
                               help='Filename that contains Constraints for the record types in the module')
    search_parser.add_argument('--extended', action='store_true',
                               help='Do not restrict the search to allocated chunks')
    search_parser.add_argument('--hint', type=argparse_utils.int16,
                               help='Restrict the search to the memory page containing this hint address')
    search_parser.set_defaults(func=search_cmdline)
    return search_parser


def show_argparser(show_parser):
    """ Show function options argument parser """
    show_parser.add_argument('record_type_name', type=str,
                             help='Python record type name. Module must be in Python path')
    show_parser.add_argument('address', type=argparse_utils.int16, help='Record memory address in hex')
    show_parser.add_argument('--constraints_file', type=argparse.FileType('r'),
                             help='Filename that contains Constraints for the record types in the module. '
                             'The validation results will be shown on stdout.')
    show_parser.set_defaults(func=show_cmdline)
    return show_parser


def watch_argparser(watch_parser):
    """ Watch function options argument parser """
    # only useful for live PID. Not rekall/vol.
    watch_parser.add_argument('record_type_name', type=str,
                              help='Python record type name. Module must be in Python path')
    watch_parser.add_argument('address', type=argparse_utils.int16, help='Structure memory address')
    watch_parser.add_argument('refresh_rate', type=int, default=0, help='Seconds between refresh')
    watch_parser.add_argument('varname', type=str, default=None,
                              help='structure member name (eg. pointername.valuename)')
    watch_parser.set_defaults(func=watch)
    return watch_parser


def dump_argparser(dump_parser):
    """ Dumper function options argument parser """
    # FIXME create the rekall/vol dumpers.
    # only useful for live PID. Not rekall/vol.
    dump_parser.add_argument('output_folder_name', type=str, help='Output to this memory dump folder')
    dump_parser.set_defaults(func=dump_process)
    return dump_parser


def output_argparser(rootparser):
    """ Output choices options argument parser """
    output = rootparser.add_mutually_exclusive_group(required=False)
    output.add_argument('--string', dest='output', action='store_const', const='string',
                        help='Print results as human readable string')
    output.add_argument('--json', dest='output', action='store_const', const='json',
                        help='Print results as json readable string')
    # useful in interactive mode
    output.add_argument('--python', dest='output', action='store_const', const='python',
                        help='Print results as python code')
    output.add_argument('--pickled', dest='output', action='store_const', const='pickled',
                        help='Print results as pickled string')
    output.set_defaults(output='string')
    return rootparser


def set_logging_level(opts):
    level = logging.INFO
    if opts.debug:
        level = logging.DEBUG
    elif opts.quiet:
        level = logging.ERROR
    #
    if opts.debug:
        flog = os.path.normpath('log')
        # FORMAT = '%(relativeCreated)d %(message)s'
        # logging.basicConfig(format=FORMAT, level=level, filename=flog, filemode='w')
        logging.basicConfig(level=level, filename=flog, filemode='w')
        print('[+] **** COMPLETE debug log to %s' % flog)
    else:
        logging.basicConfig(level=level)
    # 2.6, 2.7 compat
    sh = logging.StreamHandler(sys.stdout)
    logging.getLogger('haystack').addHandler(sh)
    return


def live_watch():
    argv = sys.argv[1:]
    desc = WATCH_DESC + DUMPTYPE_LIVE_DESC
    rootparser = base_argparser(program_name=os.path.basename(sys.argv[0]), description=desc)
    rootparser.add_argument('pid', type=int, help='Target PID on the local system')
    watch_argparser(rootparser)
    output_argparser(rootparser)
    opts = rootparser.parse_args(argv)
    opts.dumptype = DUMPTYPE_LIVE
    # apply verbosity
    set_logging_level(opts)
    # execute function
    opts.func(opts)
    return


def volatility_dump():
    argv = sys.argv[1:]
    desc = DUMP_DESC + DUMPTYPE_VOL_DESC
    rootparser = base_argparser(program_name=os.path.basename(sys.argv[0]), description=desc)
    rootparser.add_argument('dump_filename', type=argparse_utils.readable, help='Use this memory dump file')
    rootparser.add_argument('pid', type=int, help='Target PID in the OS memory dump')
    dump_argparser(rootparser)
    opts = rootparser.parse_args(argv)
    opts.dumptype = DUMPTYPE_VOLATILITY
    # apply verbosity
    set_logging_level(opts)
    # execute function
    opts.func(opts)
    return


def rekall_dump():
    argv = sys.argv[1:]
    desc = DUMP_DESC + DUMPTYPE_REKALL_DESC
    rootparser = base_argparser(program_name=os.path.basename(sys.argv[0]), description=desc)
    rootparser.add_argument('dump_filename', type=argparse_utils.readable, help='Use this memory dump file')
    rootparser.add_argument('pid', type=int, help='Target PID in the OS memory dump')
    dump_argparser(rootparser)
    opts = rootparser.parse_args(argv)
    opts.dumptype = DUMPTYPE_REKALL
    # apply verbosity
    set_logging_level(opts)
    # execute function
    opts.func(opts)
    return


def search():
    argv = sys.argv[1:]
    desc = SEARCH_DESC
    rootparser = base_argparser(program_name=os.path.basename(sys.argv[0]), description=desc)
    search_argparser(rootparser)
    output_argparser(rootparser)
    opts = rootparser.parse_args(argv)
    # apply verbosity
    set_logging_level(opts)
    # execute function
    opts.func(opts)
    return


def show():
    argv = sys.argv[1:]
    desc = SHOW_DESC
    rootparser = base_argparser(program_name=os.path.basename(sys.argv[0]), description=desc)
    show_argparser(rootparser)
    output_argparser(rootparser)
    opts = rootparser.parse_args(argv)
    # apply verbosity
    set_logging_level(opts)
    # execute function
    opts.func(opts)
    return


if '__main__' == __name__:
    search()
